proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -467,29 +478,31 @@ lemma st_tcb_at_kh_if_split:
         else st_tcb_at_kh P ptr kh)"
   by (fastforce simp: st_tcb_at_kh_def obj_at_kh_def obj_at_def)
 
-lemma set_thread_state_valid_etcbs[wp]:
-  "\<lbrace>valid_etcbs\<rbrace> set_thread_state ref ts \<lbrace>\<lambda>_. valid_etcbs\<rbrace>"
-  apply (simp add: set_thread_state_def)
-  apply (wp | wpc | simp add: set_thread_state_ext_def set_object_def get_object_def)+
-  apply (fastforce simp: valid_etcbs_def st_tcb_at_kh_if_split
-                   dest: get_tcb_st_tcb_at)
-  done
+crunch set_thread_state_act
+  for valid_queues[wp]: valid_queues
 
-lemma set_bound_notification_valid_etcbs[wp]:
-  "\<lbrace>valid_etcbs\<rbrace> set_bound_notification ref ntfn \<lbrace>\<lambda>_. valid_etcbs\<rbrace>"
-  apply (simp add: set_bound_notification_def)
-  apply (wp | wpc | simp add: set_object_def get_object_def)+
-  apply (fastforce simp: valid_etcbs_def st_tcb_at_kh_if_split
-                   dest: get_tcb_st_tcb_at)
-  done
+lemma etcbs_of_update_unrelated:
+  "\<lbrakk>kh ref = Some (TCB tcb); etcb_of tcb = etcb_of tcb'\<rbrakk> \<Longrightarrow>
+    etcbs_of' (\<lambda>r. if r = ref then Some (TCB tcb') else kh r) = etcbs_of' kh"
+  by (auto simp: etcbs_of'_def)
+
+lemma etcbs_of_update_state[simp]:
+  "get_tcb ref s = Some tcb \<Longrightarrow>
+   etcbs_of' (\<lambda>r. if r = ref then Some (TCB (tcb_state_update f tcb)) else kheap s r) = etcbs_of' (kheap s)"
+  by (auto simp: etcbs_of_update_unrelated dest!: get_tcb_SomeD)
```

Comment:
this doesn't look like a good candidate for `[simp]`, it both has a conditional and a *very* specific pattern to match on, that looks like an unfolded fun_upd; I remember on AARCH64 proofs we had issues like that with fun_upd somewhere, but don't remember where. I think we stuck with the fun_upd syntax though and prevented unfolding somewhere

Reply:
Hmm, how against this sort of lemma are you? It was copied from rt and that specific pattern comes up in a few places in the proofs, so I don't really want to try and rework things to avoid it. I did have a quick look and the fun_upd seems to be expanded by `fun_upd_apply`, which would be a pain to have to delete from `simp` everywhere.

Reply:
If they're copied, I think for this PR we live with them.
I'm not sure about the fun_upd situation precisely, but one way I would handle it is writing the lemma with a `:=` so that it makes sense, and using `[simplified fun_upd_def]` (or `fun_upd_apply` or whatever, to get it into simp normal form. Generally it looks really noisy, but I guess that's fun_upd for you.

Reply:
There's definitely a fair bit of `supply fun_upd_apply[simp del]` and `simp_del: fun_upd_apply` on AARCH64

Reply:
The lemma is not pretty, but it should be safe in the sense that the lhs of the equation is specific enough for Isabelle not to go chasing after the side condition all the time. So it shouldn't slow things down.

And agreed that this smells like fun_upd_apply repair which should ideally prevented before it happens. Ok to live with it for now, I'd say.

Reply:
It's of course not just `fun_upd_apply` repair, it is actual content: it's showing that a `tcb_state_update` is invisible in the `etcbs_of` projection. That part is actually needed. The phrasing as an expanded `fun_upd_apply` is not so great, but if it is the actual normal form in these proofs, then it's not unjustified, just a bit ugly.